llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,215 @@
+
+  // dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << CurrentLoop.getHeader()->getName() << "\n");
+
+  if (!RawMinIterationCount.has_value() || RawMinIterationCount.value() <= 0)
+    return PreservedAnalyses::all();
+  int MinIterCount = RawMinIterationCount.value();
+
+  if (!isLoopRotateCandidate(CurrentLoop)) {
+    LLVM_DEBUG(
+        dbgs()
+        << "LoopIterCountsToAssumptions-Warning: Loop "
+        << CurrentLoop.getHeader()->getParent()->getName() << " "
+        << CurrentLoop.getName()
+        << " will not be rotated, therefore aborting adding assumptions.\n");
+    return PreservedAnalyses::all();
+  }
+
+  LLVM_DEBUG(CurrentLoop.getHeader()->getParent()->dump(););
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << CurrentLoop.getHeader()->getParent()->getName() << " "
+                    << CurrentLoop.getName() << " (" << MinIterCount << ")\n");
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return PreservedAnalyses::all();
+
+  // check if the branch instruction is the correct one
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI || BI->isUnconditional())
+    return PreservedAnalyses::all();
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return PreservedAnalyses::all();
+
+  LLVM_DEBUG(dbgs() << "Branch Instruction Found: "; LoopCmpInstr->dump();
+             dbgs() << "\n");
+
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountsToAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return PreservedAnalyses::all();
+  };
+
+  // get Scalar Evolution of the induction variable
+  unsigned IVOperand = -1;
+  const SCEVAddRecExpr *AddRecExpr =
+      getLoopCmpIVAddRecExpr(*LoopCmpInstr, SE, IVOperand);
+  if (!AddRecExpr) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountsToAssumptions-Warning: Could not extract "
+                  "SCEVAddRecExpr! Will not process Metadata\n");
+    return PreservedAnalyses::all();
+  }
+
+  LLVM_DEBUG(dbgs() << "Loop Metadata        : " << MinIterCount << "\n");
+
+  IRBuilder<> Builder(CurrentLoop.getHeader()->getTerminator());
+
+  SCEVExpander Expander(
+      *SE, CurrentLoop.getHeader()->getParent()->getParent()->getDataLayout(),
+      "expanded");
+
+  // get Minimum Iteration Value Expression
+  const SCEV *MinIterSCEV = AddRecExpr->evaluateAtIteration(
+      SE->getConstant(APInt(32, MinIterCount)), *SE);
+
+  // in the case of decrementing, add NSW flag, so that the expression can be
+  // simplified
+  if (auto *AE = dyn_cast<SCEVAddExpr>(MinIterSCEV)) {
```

Comment:
Checking: Even after evaluating the expression at a specific iteration, it remains a `SCEVAddExpr`? So essentially this just offsets the start value?

Reply:
no, it can also reduce to a constant, in the case that the Start point is a constant, i.e. if the below loop has an iteration count of 5, SCEV will provide the constant, since it can be evaluated.

`for (int i=0, i < N; i++) `